contracts/cw20-ics20/src/amount.rs

```diff
@@ -64,3 +66,158 @@ impl Amount {
+            ));
+        }
+
+        // mimicks formats like hex or bech32 where different casings are valid for one address
+        let normalized = input.to_lowercase();
+
+        let mut out = Vec::from(normalized);
+
+        // pad to canonical length with NULL bytes
+        out.resize(api.length_max, 0x00);
+        // content-dependent rotate followed by shuffle to destroy
+        // the most obvious structure (https://github.com/CosmWasm/cosmwasm/issues/552)
+        let rotate_by = Self::digit_sum(&out) % api.length_max;
+        out.rotate_left(rotate_by);
+        for _ in 0..api.shuffles_encode {
+            out = Self::riffle_shuffle(&out);
+        }
+        Ok(out.into())
+    }
+
+    fn addr_humanize(&self, canonical: &CanonicalAddr) -> StdResult<Addr> {
+        let api = Self::default();
+
+        if canonical.len() != api.length_max {
+            return Err(StdError::generic_err(
+                "Invalid input: canonical address length not correct",
+            ));
+        }
+
+        let mut tmp: Vec<u8> = canonical.clone().into();
+        // Shuffle two more times which restored the original value (24 elements are back to original after 20 rounds)
+        for _ in 0..api.shuffles_decode {
+            tmp = Self::riffle_shuffle(&tmp);
+        }
+        // Rotate back
+        let rotate_by = Self::digit_sum(&tmp) % api.length_max;
+        tmp.rotate_right(rotate_by);
+        // Remove NULL bytes (i.e. the padding)
+        let trimmed = tmp.into_iter().filter(|&x| x != 0x00).collect();
+        // decode UTF-8 bytes into string
+        let human = String::from_utf8(trimmed)?;
+        Ok(Addr::unchecked(human))
+    }
+
+    fn addr_validate(&self, input: &str) -> StdResult<Addr> {
+        let canonical = self.addr_canonicalize(input)?;
+        let normalized = self.addr_humanize(&canonical)?;
+        if input != normalized {
+            return Err(StdError::generic_err(
+                "Invalid input: address not normalized",
+            ));
+        }
+
+        Ok(Addr::unchecked(input))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::{Amount, Coin, Cw20CoinVerified, Uint128};
+    use cosmwasm_std::Addr;
+
+    #[test]
+    fn from_parts_native() {
+        let amount = Uint128::new(1_000_000);
+        let denom = "token-denom";
+
+        assert_eq!(
+            Amount::from_parts(denom.into(), amount),
+            Amount::Native(Coin {
+                denom: denom.into(),
+                amount
+            })
+        );
+    }
+
+    #[test]
+    fn from_parts_cw20() {
+        let amount = Uint128::new(1_000_000);
+        let cw20_addr = Addr::unchecked("token-addr");
+        let cw20_denom = "cw20:token-addr";
+
+        assert_eq!(
+            Amount::from_parts(cw20_denom.into(), amount),
+            Amount::Cw20(Cw20CoinVerified {
+                address: cw20_addr,
+                amount
+            })
+        );
+    }
+
+    #[test]
+    #[should_panic]
```

Comment:
Definitely panic is also bad, which is why changing API and returning Result on a verify method is cleaner